Fix prims.broadcast_in_dim functionalization

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #163377

Fixes #163037

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

ezyang reviewed (changes requested):
Unfortunately, it seems Codex found an existing wrong functionalization rule for rng in the prims directory. Here's what a real, full-bodied view functionalization rule looks like:

```
    at::Tensor view(c10::DispatchKeySet dispatchKeySet, const at::Tensor & self, c10::SymIntArrayRef size) {

      at::Tensor self_;
      if (at::functionalization::impl::isFunctionalTensor(self)) {

        self_ = at::functionalization::impl::from_functional_tensor(self);
      } else {
        self_ = self;
      }
      if (!at::functionalization::impl::isFunctionalTensor(self)) {
        // functionalization is re-entrant, but will no-op if it wasn't passed a FunctionalTensorWrapper.
        at::AutoDispatchSkipFunctionalize guard;
        return at::_ops::view::call(self_, size);
      }
      auto reapply_views = at::functionalization::impl::getFunctionalizationReapplyViewsTLS();
      auto inverse_return_mode = (
          reapply_views ? at::functionalization::InverseReturnMode::ViewOrScatterInverse
            : at::functionalization::InverseReturnMode::NeverView
      );
      auto compute_reference_meta =
        self.key_set().has_backend(c10::BackendComponent::XLABit) ||
        self.key_set().has_backend(c10::BackendComponent::LazyBit);
      at::Tensor reference_tensor_output;
      if (compute_reference_meta && !disable_meta_reference()) {
        auto self_meta = to_meta(self);
        at::AutoDispatchSkipFunctionalize func_guard;
        c10::impl::ExcludeDispatchKeyGuard guard(exclude_keys_for_meta_dispatch);
        reference_tensor_output = at::_ops::view::call(self_meta, size);
      }
      at::Tensor tmp_output;
      {
        at::AutoDispatchSkipFunctionalize guard;
        if (reapply_views) {
          tmp_output = at::_ops::view::call(self_, size);
        } else {
          tmp_output = at::_ops::view_copy::call(self_, size);
        }
      }

      bool has_symbolic_inputs = false;
      has_symbolic_inputs = has_symbolic_inputs | (std::any_of(size.begin(), size.end(), [=](auto& arg) { return arg.is_symbolic(); }));
      at::functionalization::ViewMeta view_meta = at::functionalization::ViewMeta(
        [reapply_views = reapply_views, size = size.vec()](const at::Tensor & base, int64_t mutated_view_idx) -> at::Tensor {
          if (reapply_views) {
            return at::_ops::view::call(base, size);
          } else {
            return at::_ops::view_copy::call(base, size);
          }
        },
        [inverse_return_mode = inverse_return_mode, size = size.vec()](const at::Tensor & base, const at::Tensor & mutated_view, int64_t mutated_view_idx) -> at::Tensor {
          return at::functionalization::FunctionalInverses::view_inverse(base, mutated_view, inverse_return_mode, size);
        },
        /*has_symbolic_inputs=*/has_symbolic_inputs,
        /*is_multi_output=*/false,
        /*is_as_strided=*/false
      );
      auto out = at::functionalization::impl::create_functional_tensor_with_view_meta(tmp_output, self, view_meta);
      // See  Note [Propagating strides in the functionalization pass]
      if (compute_reference_meta && !disable_meta_reference()) {
        at::functionalization::impl::set_sizes_strides_offset(out, reference_tensor_output);
      }
      return out;
    }

```

for broadcast_in_dim to be implemented correctly it needs to follow this structure more closely

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

laithsakka commented on torch/_prims/__init__.py:
```diff
@@ -1362,6 +1363,94 @@ def _broadcast_in_dim_aten(a, shape, broadcast_dimensions):
+    original_idx = 0
+    for idx in range(len(shape)):
+        if idx in broadcast_dims_set:
+            size = tensor_sizes[original_idx]
```
can we write this to be unbacked friendly 
the unbacked semantics for broadcasting is that if we cant tell if its a broadcast case or not
we would assume no broadcasting and do a torch_check.
example this is the meta version

https://www.internalfb.com/code/fbsource/[e18e1578407a804d7877bf6be709197b739f6eae]/fbcode/caffe2/torch/_prims/__init__.py?lines=1296

jansel committed:
Update

[ghstack-poisoned]

jansel commented:
Updated prompt (I reverted the prior version and redid it):

> Write a functionalization rule in C++ for broadcast_in_dim to fix failures in `repro_broadcast_in_dim.py` and the test test_prims_broadcast_in_dim_alias.  Try to follow the pattern of other rules like view.  Perhaps we could even rewrite broadcast_in_dim to view during functionalization.  At the end, please provide a one paragraph summary of your approach for use in a github comment.  When you change C++ rebuild with `python setup.py develop`, you have a working build environment.


Summary:

> Implemented a Functionalize kernel for prims::broadcast_in_dim that rewrites the op into the
>   same unsqueeze+expand view pattern we use in eager, captures the resulting size/stride metadata, and registers a prims-
>   specific rule so functionalization rewrites it to the appropriate as_strided view; with the new rule in place both the
>   standalone repro and CPUReproTests.test_prims_broadcast_in_dim_alias now pass

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

ngimel commented:
When would this functionalization be needed? this prim should not be used in the usual lowering process, and if someone wants to just call the function that has this behavior, they can write it with `unsqueeze` and `expand` and it will be functionalized normally.

ngimel commented:
well they shouldn't
